### ISSUE TITLE
implement cascading and variadic lambdas

### ISSUE THREAD

**Issue by porkybrain** [edouarda/brigand#186]
#177 was getting super full so I split it into greedy 'call' which is #185 and this thread.

In essence a TMP lambda is a tree of nested metafunctions, in other words a tree of types. If we start at the root unwrapping the first type we will find a lazy metafunction containing a variadic pack of types. each of these types will fit one of four catagories: a bound argument wrapped with `pin<>`, a plaeholder like `_1`, another metafunction or a lambda wrapped in `call` or  `lcall`. `pin<>` arguements are simply unwrapped, placeholders are swapped out, metafunctions are called recursively and lambdas are unwrapped and executed by their surrounding metafunctions. The inner working of this can be seen [here](https://github.com/edouarda/brigand/pull/187/commits/ea46777869db8a346233865c8f33a2379ee017ba), its actually quite simple (without all the fast tracks which will follow). 
Now with that as a foundation we still need a method of accessing the parents arguements from a child lambda when cascading. In the example `transform<l,lazy::transform<_1,lcall<myF<_1,parent<_1>>>>>` myF gets one parameter from the second transform and one from the first. In order to do this we need to pass the arguements from the first to the arguements of the second. This is a little harder than it may sound since myF is called in the guts of transform and our recursive walking algorithm cannot reach into transform directly and insert parameters. The trick is to swap out the `lcall` with our own internal wrapper called `packaged_lcall` which also contains a variadic pack of lists of parents parameters. This way our walker does not need to know anything about the algorithm (in this case transform) it just trusts that it will transparently pass through any arguments we hide in `packaged_lcall` back to our walker.

**Comment by porkybrain** [comment]
The real breaking change stems from the cost of SFINAE testing whether a type has a nested `type` alias or not (meaning it should be evaluated as a metafunction). My tests show that doing a trivial transform (using `std::is_same<>`) on 1000 lists of four elements each enabling SFINAE slows things down by about 60%, this is not negligible.

On the other hand it buys you not having to use `pin<>` and allows "type wrapping" lambdas as in `transform<T, U, std::pair<_1,_2>>`. 

It would also make this change break far less stuff (basically only code where the user wrote their own metafunctor with a nested apply template would we break stuff).

**Comment by porkybrain** [comment]
One reason I think its important to get this right is that it actually is possible to beat even well written hand coded stuff with more readable lambda code. For example because fold is fast tracked `fold<T,i<0>,call<add>>` is like 3x faster than a non fast tracked hand coded recursive add. Since no one in their right mind will be fast tracking their own user code I think it is safe to say there is potential to beat user code with lambdas in some cases. I know there are some pieces of kvasir which I hand coded for speed which are going to be turned back into lambdas, for more speed!

**Comment by ericniebler** [comment]
My thinking is to never, ever do the weird MPL look-for-nested-`::type`-and-use-it-if-it-exists thing. Rather, there should be a wrapper, maybe called `lazy::_t` or something, that means: after recursive substitution, reach in a get a nested `::type`. The default is to not do that.

**Comment by brunocodutra** [comment]
> My thinking is to never, ever do the weird MPL
look-for-nested-::type-and-use-it-if-it-exists thing.

That and the horrific treat-as-lambda-only-if-arg-is-present.

**Comment by porkybrain** [comment]
yeah that's pretty much what I was thinking, but it would break a lot of current brigands user code and make the transition from boost.MPL to brigand slightly harder. The other way to go would be add functionality to the 'call' wrapper making it the "modern" way to go and leaving current brigand stuff mostly alone.  
`transform<T,U,call<std::pair>>` should already work but more interesting would be using it with a third fixed arg or something.

**Comment by porkybrain** [comment]
essentially if we were to establish a pattern of everything on the eager side of things is a `template<template<typename...> class F, typename...Ts>` where if Ts exist they are each called with all args passed it would yield a simple pattern, simple is good. On the other hand the user would essentially have to wrap everything to achieve that pattern. We can swap out what placeholder actually are to make them fit this pattern without breaking anything. For example if you wanted to swap the order it would then be `transform<T,U,call<std::pair, _2, _1>>`. Or more interesting if we anted to do some more complex stuff `transform<T,call<std::add_pointer_t,call<std::add_pointer_t,call<std::pair,_1,_2>>>>'. 

The thing I have not figured out how to do is cascading lambdas with this model.

**Comment by porkybrain** [comment]
```C++
template<template<typename...> class F, typename...Ts>
struct c_ {};

template<typename T>
struct p_ {};

template<typename T, typename...Ts>
using a1 = T;
template<typename, typename T, typename...Ts>
using a2 = T;

using arg1 = c_<a1>;
using arg2 = c_<a2>;

template<typename T, typename...As>
struct walker {};
template<typename T, typename...As>
struct walker<p_<T>,As...> {
	using type = T;
};
template<template<typename...> class F, typename...Ts, typename...As>
struct walker<c_<F, Ts...>, As...> {
	using type = F<typename walker<Ts, As...>::type...>;
};
template<template<typename...> class F, typename...As>
struct walker<c_<F>, As...> {
	using type = F<As...>;
};

using r1 = typename walker<c_<std::add_pointer_t, c_<std::add_pointer_t, c_<std::pair, arg2, arg1>>>, int, bool>::type;

using r2 = typename walker<c_<std::add_pointer_t, c_<std::add_pointer_t, c_<std::tuple, arg2, arg1, p_<long>>>>, int, bool>::type;
```
this is roughly how that would work.

**Comment by brunocodutra** [comment]
> essentially if we were to establish a pattern of everything on the eager
side of things is a template<template<typename...> class F, typename...Ts>
where if Ts exist they are each called with all args passed it would yield
a simple pattern, simple is good.

The problem with this pattern is that it is not symmetrical. If one thinks
of it as semantic synonym for bind<quote<F>, Ts...>, one is tempted to
believe zero Ts... means F is bound to zero args, but, instead, in your
scheme it means F is not bound at all.

**Comment by porkybrain** [comment]
Im benchmarking nested stuff, how would I write `std::add_pointer<std::add_pointer<std::pair<_1, _1>>>` in meta and metal most efficiently?

**Comment by porkybrain** [comment]
Perliminary test shows using wrapped aliases to select the correct arg rather than a specialization of the walker template is faster (1000 transforms of unique four element lists):

walker with alias selector: 1.88 s 
walker with specialization selector: 1.98 s 
new brigand with SFINAE 6.8 s
old brigand with SFINAE 8.8 s

The fact that the alias was faster than the specialization is interesting, maybe I could build one where p_ is also just a special c_ and use specialization only to unpack things.

**Comment by brunocodutra** [comment]
> Im benchmarking nested stuff, how would I write `transform<T, std::add_pointer<std::add_pointer<std::pair<_1, _1>>>>` in meta and metal most efficiently?

Using Metal@bare it should look like the following

~~~{.cpp}
transform<
    bind<
        lambda<std::add_pointer_t>, 
        bind<
            lambda<std::add_pointer_t>,
            bind<lambda<std::pair>, _1, _1>
        >
    >, 
    T
>
~~~

1. Yes, it is verbose because I want to trim the fat before bulking up again, so I haven't provided the lazy adaptors yet;
2. Notice the Lambda comes first in Metal's `transform`, becuase it is actually variadic and one should be able to provide an arbitrary number of Lists, provided of course they are all of the same length and the Lambda's arity is respected.

**Comment by porkybrain** [comment]
I made a hybrid of the old brigand and the new call syntax [here](https://github.com/edouarda/brigand/pull/187/commits/6e5ceae40a66248a3c3cb1547d205512c6fb1067)

Essentially if you want to call something lazily then just drop in placeholders, if you want to call something eager use `call<thing, args, or, placeholders>`, you can also interchange them. You can also use `lcall<callable_something>` to defer the execution of that thing (maybe we should rename it). You can use `parent<_1>` inside deferred lambdas so the parents args are accessible in an equally powerful way as in meta's cascading lambdas. Sadly although `pin<>` is not needed inside `call` statements it is needed inside lazy evaluated statements if the pinned value is a template with all type parameters.

So how about speed? (take with grain of salt, my machine transforming 1000 unique 4 element list)
```
old brigand: 8s
transform<T, std::add_pointer<std::add_pointer<std::pair<_1, _1>>>>;		

new brigand SFINAE: 6.80s (this is what removing pin and SFINAE testing would cost us)
transform<T, std::add_pointer<std::add_pointer<std::pair<_1, _1>>>>;	

newest brigand pure call: 1.77s
transform<T, call<std::add_pointer_t, call<std::add_pointer_t, call<std::pair, _1, _1>>>>;

newest brigand mixed call: 1.76s
transform<T, std::add_pointer<std::add_pointer<call<std::pair,_1, _1>>>>;					

metal: 2.39s
transform<bind<lambda<std::add_pointer_t>,bind<lambda<std::add_pointer_t>,bind<lambda<std::pair>, _1, _1>>>,T>;

How about shorter lambdas:
meta: 1.17s
transform < T, quote<std::add_pointer_t> >;  

metal: 1.15s
transform<bind<lambda<std::add_pointer_t>, _1>, T>

brigand call: 0.94s (0.68s with fast track)
transform<T, call<std::add_pointer_t, _1>>;	

brigand lazy exact same numbers 
transform<T, std::add_pointer<_1>>;	
```
In conclusion I think a unified syntax allowing both eager and lazy calls and cascading can still beat all current implementations except @pdimov style which, by definition, is impossible.

**Comment by brunocodutra** [comment]
Nice work!

I'm happy Metal could actually keep up without such fancy techniques. 

BTW, `transform<bind<lambda<std::add_pointer_t>, _1>, T>` is equivalent to `transform<lambda<std::add_pointer_t>, T>`, so **maybe** Brigand isn't **that** faster for shorter lambdas.

**Comment by porkybrain** [comment]
thanks for the correction, I corrected the numbers and added MSVC. Brigand also supports variadic transform but still with the lambda at the back. looks like metal is actually pretty equal with trivial lambdas.

**Comment by brunocodutra** [comment]
Nice! This is the first time I see benchmark results for Metal on MSVC btw, and [not so surprisingly at all] it seems it chokes real bad on `bind`. Well such is life, MSVC sucks and there's not much one can do about it (besides ditching C1 in favour of Clang).

<sub>* I really hope `60s` is actually a typo ;)</sub>

**Comment by porkybrain** [comment]
> * I really hope 60s is actually a typo ;)
lol fixed.

meta doesn't even compile on MSVC15 (I am missing the newest update). I'm not really interested about MSVC but I understand there are a few who need it.

**Comment by brunocodutra** [comment]
> I'm not really interested about MSVC but I understand there are a few who need it. 

Indeed, but It's so annoying having to spend countless hours working around its *imperfections* just to get it to work ¬¬

**Comment by edouarda** [comment]
> meta doesn't even compile on MSVC15 (I am missing the newest update). I'm not really interested about MSVC but I understand there are a few who need it.

It's not a few, it's a lot of people, including me! Windows is still a major development platform.

**Comment by edouarda** [comment]
If I understand correctly, brigand::call is a bind?

**Comment by porkybrain** [comment]
yeah `call` is probably `bind` and `lcall` is probably `defer` or `protect`, I suck at naming things. What do you guys think, would this be the correct naming convention?

```C++
template<typename T>
reverse_fold<T,list<list<>>,
    bind<join,bind<transform, _2,
        defer<bind<join,bind<transform,parent<_1>,
            defer<bind<list,bind<push_front<_1,parent<_1>>>>>>>>>>;

//or
namespace l = brigand::lazy;
template<typename T>
reverse_fold<T,list<list<>>,
    l::join<l::transform<_2,
        defer<l::join<l::transform<parent<_1>,
            defer<l::list<l::push_front<_1,parent<_1>>>>>>>>>>;

//as opposed to

namespace l = meta::lazy;
template <class L>
using cart_prod = reverse_fold<L, list<list<>>,
    lambda< _a, _b, l::join<l::transform< _b, 
        lambda<_c, l::join<l::transform<_a, 
            lambda<_d, list<l::push_front<_d, _c>>>>>>>>>>;
```

**Comment by porkybrain** [comment]
Another interesting benchmark, meta's bind was the one place I thought we might have trouble beating:
```C++
meta with bind: 0.72s
meta::transform < T, meta::bind_front<meta::quote<std::pair>,int> >;

brigand: 0.47s
transform<T, call<std::pair, int, _1>>;
```

**Comment by edouarda** [comment]
I think bind is easier for people to grasp, especially if they come from Boost.MPL.

**Comment by porkybrain** [comment]
I think I agree on bind, what should we call defer/protect/lambda/lcall? Basically its the thing you wrap a lambda with if you don't want it to be called in current context but do want it to be called by its surrounding algorithm as in `transform<T,l::transform<_1,___this_protector_thing___<add_pointer<_1>>>'?

**Comment by edouarda** [comment]
defer sounds good to me. @jfalcou any opinion?

**Comment by porkybrain** [comment]
[inherit_linearly.cpp](https://github.com/edouarda/brigand/blob/e9374b3450c924f706e2f625daab7780296180de/test/inherit_linearly.cpp) is giving me problems, as far as I can see the user is putting `brigand::_element` and `brigand::_state`, which are really just `_1` and `_2` into their own type. This syntax however cannot be supported as a lambda unless we use SFINAE on lambdas. 
I could pattern match and turn `f<args..>` into `bind<f,args...>` under the hood but seems like kind of a hack. What do you guys think? I'm almost done with propagating the changes throughout brigand. Actually this lambda change gets rid of a whole lot of stuff like placeholder tests and quote and such.

**Comment by porkybrain** [comment]
moving conversation to the pull request #187
